### PR TITLE
Migrates Job Scheduler plugin to OpenSearch modules

### PR DESCRIPTION
**Pull request by dbbaughe**
### Description
This PR migrates the Job Scheduler plugin to OpenSearch modules.

Would be a massive task to attempt to PR a single commit of this so the way its broken down is:

1. First commit purely takes the job scheduler plugin from a sibling directory and moves it under modules/job-scheduler in core minus the .git directory.
2. Second commit removes a bunch of files, scripts, directories, etc. relating to job scheduler as a plugin.
3. Third commit refactors the build.gradle files for job scheduler and job scheduler SPI
4. Fourth commit adds the modules to the missing javadocs task as this was focused on migration.
5. Fifth commit moves the sample plugin to the plugins/examples directory and renames it/fixes up file/package paths.
6. Sixth commit calls the style linter to fix style issues.
7. Seventh commit adds the cron-utils dependency license files to the job scheduler spi license directory
8. Eighth commit fixes the breaking changes in job scheduler from the 2.0 upgrade
9. Ninth commit adds a plugin security policy
10. Rest of commits are some small styling/migration fixes

Unless we want to review the entire job scheduler plugin again, I'd recommend looking at the individual commits to see the work that was done to integrate the job scheduler plugin into the core repository.
 
### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/2534
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
I'd like to revive the conversation about merging this PR, and hear a strong opinion from @nknize @mch2 @andrross @kartg @reta and anyone else who cares to comment. Do we want this change in core?

File: modules/job-scheduler/src/main/java/org/opensearch/jobscheduler/sweeper/JobSweeper.java [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+        this.xContentRegistry = registry;
+        this.indexToProviders = indexToProviders;
+        this.scheduler = scheduler;
+        this.lockService = lockService;
+
+        this.lastFullSweepTimeNano = System.nanoTime();
+        this.loadSettings(settings);
+        this.addConfigListeners();
+
+        this.fullSweepExecutor = Executors.newSingleThreadExecutor(OpenSearchExecutors.daemonThreadFactory("opendistro_job_sweeper"));
```
Comment (reta) [inline modules/job-scheduler/src/main/java/org/opensearch/jobscheduler/sweeper/JobSweeper.java:117]: Minor `opendistro_job_sweeper` -> `opensearch_job_sweeper`?

File: modules/job-scheduler/src/test/java/org/opensearch/jobscheduler/JobSchedulerSettingsTests.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    public void testSettingsGetValue() {
+        Settings settings = Settings.builder().put("plugins.jobscheduler.request_timeout", "42s").build();
+        assertEquals(JobSchedulerSettings.REQUEST_TIMEOUT.get(settings), TimeValue.timeValueSeconds(42));
+        assertEquals(LegacyOpenDistroJobSchedulerSettings.REQUEST_TIMEOUT.get(settings), TimeValue.timeValueSeconds(10));
+    }
+
+    public void testSettingsGetValueWithLegacyFallback() {
+        Settings settings = Settings.builder()
+            .put("opendistro.jobscheduler.request_timeout", "1s")
```
Comment (reta) [inline modules/job-scheduler/src/test/java/org/opensearch/jobscheduler/JobSchedulerSettingsTests.java:76]: Should we rename the setting for 2.0 to be `opensearch` instead?

**Comment by kartg** [comment]
A lot of the for/against discussion can be viewed in https://github.com/opensearch-project/job-scheduler/issues/147

IMO, this change should _not_ go forward because "job scheduling" is not core functionality in a distributed search engine. The primary purpose of this feature is to be a building block for other plugins - on its own, it does not add any value. Since plugins are by nature optional, this feature cannot be considered essential.

The pain point driving this PR is that version upgrades are slowed because plugins are dependent on Job Scheduler upgrading before they can upgrade. In addition to seeing data to quantify "slow", I think we should consider tackling this with a different approach - given that Job Scheduler does not provide value on its own, is it really a _plugin_? 

Can we decouple Job Scheduler enough so that it is not tied to core versions? Is it better modeled as a _library_ that other plugins import?

**Comment by nknize** [comment]
I lost track of this PR due to so many other in flight issues. I wanted to circle back here and make sure this didn't go stale for much longer than it already has. 

Personally, I don't have a problem moving this to the core repo if it makes sense. I think I'm missing a few key answers:

1. Is there a one to many relationship between this function and other plugins? (e.g., how many plugins depend on this as a core function today?)
2. Do we forsee this as an enabler for many other plugins?
3. Does it have to be a module? I think users might want the option to install this function only when needed. 

To me, one criteria worthy of considering for promotion of a plugin from "second class" (external repo) to "first class" (core repo) is the number of plugins that depend on the plugin for base functionality. I don't think we should necessarily limit the entire core repository to "search engine" only functions; `/server` directory sure, but the modules, plugins, and libraries directories are specifically designed for situations like this to help minimize the "Dependency spaghetti" between external repos. If job scheduler is already a one to many dependency, then moving this to the core repo makes sense. We just need to be thoughtful about whether it's a module or a plugin and for that I always err on the side of plugin so we don't force users to install a plugin that consumers resources they'll never use.

**Comment by Bukhtawar** [comment]
I would second @nknize thoughts on enabling simpler dependency management as another criteria for core plugins, which likely is the case with job-scheduler. I think we should also have this clearly documented as a guidance for adding future module/plugins to core. 

Also do we really need the complete implementation of job-scheduler to be in core, can we just have the abstract layers in core to simplify dependency management and have the actual implementation sitting outside it?

**Comment by praveensameneni** [comment]
> 1. Is there a one to many relationship between this function and other plugins? (e.g., how many plugins depend on this as a core function today?)

Yes, there is a 1:M relationship. We have Index-Management, Anomaly Detection, Reporting using Job Scheduler. Alerting has an inbuilt Job Scheduler, which was the genesis of Job Scheduler as we saw the use case as a more generic one.

> 2\. Do we forsee this as an enabler for many other plugins?

It can potentially be one - e.g Observability and Security Analytics will have use cases to get traces / security findings based on a time interval.

> 3\. Does it have to be a module? I think users might want the option to install this function only when needed.

Open to being it under plugins directory. 

> To me, one criteria worthy of considering for promotion of a plugin from "second class" (external repo) to "first class" (core repo) is the number of plugins that depend on the plugin for base functionality. I don't think we should necessarily limit the entire core repository to "search engine" only functions; `/server` directory sure, but the modules, plugins, and libraries directories are specifically designed for situations like this to help minimize the "Dependency spaghetti" between external repos. If job scheduler is already a one to many dependency, then moving this to the core repo makes sense.

Couldn't agree more!

**Comment by praveensameneni** [comment]
@kartg , @CEHENKLE , @dblock , Do you have any further questions on merging this PR?

**Comment by owaiskazi19** [comment]
Job Scheduler currently uses Service Provider Interface(SPI) design to interact with plugins like Anomaly Detection, Index-Management. If we are deciding to move Job Scheduler to core should the design still remain the same for interactions of the plugins with Job Scheduler and what impact will this design have on the latency?

**Comment by minalsha** [comment]
Keeping this PR open to keep it for reference once we are able to continue the work to move JS into native plugins

**Comment by minalsha** [comment]
@dreamer-89  we can close this PR in lieu of https://github.com/opensearch-project/OpenSearch/pull/5453 since we would want to move JS as a native plugin so that it can be installed optionally. Will let @joshpalis  confirm. Thanks

**Comment by joshpalis** [comment]
As the decision to relocate job scheduler into core as a native plugin (optionally installed) rather than a module (installed by default) has been finalized, closing this PR. An update on the effort to relocate job scheduler into core can be found [here](https://github.com/opensearch-project/OpenSearch/issues/5310#issuecomment-1597518680)

**Pull request closed without merging**